### DISCUSSION TITLE
[Enhancement] Built in controls

### DISCUSSION THREAD

**Discussion by luismts** [dotnet/maui#93] [General]
# Summary

Within the Xamarin Forms ecosystem (community and third party companies), we have an immense amount of controls and control variations. It would be great to include controls like:

- Camera view
- Calendar
- PDF Viewer
- PopUp Page
- Charts or Gauges
- Segmented control
- Etc...

On the other hand, it would be nice to have some general functionalities integrated like

- Autocomplete View
- Rating
- Parallax View
- Treemap
- Segmented Control
- Etc..

And some built-in animations for controls and pages would be loved by the community

- Page Transitions
- Specific Controls Animations
- Visual (like Cupertino or another one)
- Etc..

Some outstanding features like

- Print
- Offline character recognition
- Digital signature
- Etc...

There are many things that can be integrated for ease of use for the entire community. There are some solutions for these controls out there, but sometimes they need a push and I think this is the best way because we will make the framework as it is much more mature in relation to what it can encompass.

**Comment by KurdCsharper** [comment]
Duplicate of #76

**Comment by charlesroddie** [comment]
@KurdCsharper not a duplicate as that post is about desires for things in general, while this is a desire for a particular class of things!

**Comment by charlesroddie** [comment]
There are some tradeoffs here.

By being in separate packages/repos, some of these controls can:
1. Be changed more easily. It's very hard to change things in the Xamarin.Forms repo and we don't yet know that MAUI will be better. E.g. Rg.Plugins.Popup is already decent and it's unclear that creating a new one in MAUI will be better.
2. Support other .Net UI systems in addition to MAUI. E.g. charts or pdf could be done using platform-independent renderers so they can be used by MAUI projects and also by UWP, WPF, MAUI.iOS/Android, Blazor projects.

That said, things that involve a lot of user interaction and use of native controls would be good candidates: Camera, Calendar, Autocomplete

Print and OCR aren't controls so I would create a separate issue for these. And transitions.

**Comment by davidortinau** [comment]
Camera we have a PR for but it was failing some tests. And I think we might be waiting for the new Google AndroidX version.

Print - https://github.com/xamarin/Essentials/issues/718

PopUp is something we should have worked on later this year if things go as planned.

SegmentedControl - Hussain has a work in progress for https://github.com/xamarin/Xamarin.Forms/issues/6904 though perhaps we considering using TabView from @jsuarezruiz as an xplat solution.

AppBar offers a parallax type of scroll and is coming, but more fine support for parallax is perhaps a scroll view enhancement. How might a Parallax control look if it were standalone? Akin to the Expander?

Page Transitions are coming with Segues. It was on the Xamarin.Forms roadmap and will be moved here.

Controls I don't expect we will do and which have great vendor support:
* Calendar
* PDF Viewer
* Charts and Graphs and Gauges
* Digital signature

Could you explain more what you see for these?
* Treemap
* Rating 
* Specific control animations

If we put/accepted some of these controls into a community toolkit, would that be useful?

What is preventing you (anyone) from using existing controls available from vendors or open source offerings today?

**Reply by luismts** [reply]
Controls like the Rating one are used in must public apps and animations in Xamarin.Forms can be really hard for beginners. Allowing developers to do so with a simple API and built-in animation package would be a really big upgrated, belive me. 

Now, and most of the time, developers are comparing tools with other ones. So, based on what others competitors/tools offer the xamarin team need to be wise about what they do and be innovative in the way they do it.

And I think that some times you can **kill two birds with one shot**.

For example: When the xamarin team lauched Xamarin.Forms.Visual Material I writted a post saying that this library would make a big achievement for the community and it was, but to be honest it is so limited based on what the  material.io documentation says.

Now, the community is asking you for FluentUI  (never used it before but I really like it based on what I have seen, I will use it with my team if you will added) because lot os things like: 

- It is really good looking UI framework
- It is from Microsoft
- Microsoft made it available for React Navite and not xamarin
- And so on

Adding support for this framework in Xamarin.Forms.Visual you can get many benefits and make a really good marketing about it promoting microsoft UI framework in must Xamarin Apps.

On the other hand, there is nothing preventing me from using existing open source controls today. But some clients don't like that vendors are tracking their apps cause they have really hard policies and that can not be discussed.

Another thing I'm seeing and some community members comment is that the general communuty is geeting monotonous talking about the same thing over and over. Thanks to MAUI that changed that, but that is for another thread.

> Sorry for my late on this.

**Reply by GiampaoloGabba** [reply]
Imho more themes need to be bundled in core (at least the fluent one... come one! Is available even for React and not for MAUI???).

For other controls i like the way they are shaping: in external packages supported by microsoft/microsoft employees/community.
The development is faster compared to the core framework, the quality is high and the support is top notch.
I have a feeling that in the next couple of months these project will be the base for every xamarin/maui project out there:

https://github.com/xamarin/Essentials
https://github.com/xamarin/XamarinCommunityToolkit
https://github.com/jsuarezruiz/TemplateUI

**Comment by ghost** [comment]
Calendar
PDF Viewer
Charts and Graphs and Gauges
Digital signature

If these controls make the size issue or the rendering issue of the app , then why can't we make a standalone plugin like essential for advanced ui controls

**Reply by davidbuckleyni** [reply]
Many PDF Components work and why is navigation to a web browser that supports PDF not suffiencet if its a secure PDF can always password as commented above I dont think they take time in those controls. Maui Needs to be slim

**Comment by mrwcjoughin** [comment]
There already is something like that :
https://github.com/xamarin/XamarinCommunityToolkit


On Fri, 25 Sep 2020 at 19:48, NithinKJoy <notifications@github.com> wrote:

>
>
> Calendar
>
>
> PDF Viewer
>
>
> Charts and Graphs and Gauges
>
>
> Digital signature
>
>
> If these controls make the size issue or the rendering issue of the app ,
> then why can't we make a standalone plugin like essential for advanced ui
> controls
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/maui/discussions/93#discussioncomment-85168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGERXGJT4ZALJX7NJEJ7TSHTJYTANCNFSM4RZWIWZA>
> .
>
>
> --
Thanks,

Matthew Joughin
Founder / CTO

+27 81 529 7169
http://www.adaptableapps.net
